src/spooled.rs

```diff
@@ -0,0 +1,136 @@
+use std::fs::File;
+use std::io::{self, Read, Write, Seek, SeekFrom, Cursor};
+use file::tempfile;
+
+/// An object that behaves like a regular temporary file, but keeps data in
+/// memory until it reaches a configured size, at which point the data is
+/// written to a temporary file on disk, and further operations use the file
+/// on disk.
+#[derive(Debug)]
+pub struct SpooledTempFile {
+    max_size: usize,
+    cursor: Option<Cursor<Vec<u8>>>,
+    file: Option<File>,
+}
+
+/// Create a new spooled temporary file.
+///
+/// # Security
+///
+/// This variant is secure/reliable in the presence of a pathological temporary
+/// file cleaner.
+///
+/// # Resource Leaking
+///
+/// The temporary file will be automatically removed by the OS when the last
+/// handle to it is closed. This doesn't rely on Rust destructors being run, so
+/// will (almost) never fail to clean up the temporary file.
+///
+/// # Examples
+///
+/// ```
+/// # extern crate tempfile;
+/// use tempfile::spooled_tempfile;
+/// use std::io::{self, Write};
+///
+/// # fn main() {
+/// #     if let Err(_) = run() {
+/// #         ::std::process::exit(1);
+/// #     }
+/// # }
+/// # fn run() -> Result<(), io::Error> {
+/// let mut file = spooled_tempfile(15);
+///
+/// writeln!(file, "short line")?;
+/// assert!(!file.rolled_over());
+///
+/// // as a result of this write call, the size of the data will exceed
+/// // `max_size` (15), so it will be written to a temporary file on disk,
+/// // and the in-memory buffer will be dropped
+/// writeln!(file, "marvin gardens")?;
+/// assert!(file.rolled_over());
+///
+/// # Ok(())
+/// # }
+/// ```
+pub fn spooled_tempfile(max_size: usize) -> SpooledTempFile {
+    SpooledTempFile {
+        max_size: max_size,
+        cursor: Some(Cursor::new(Vec::new())),
+        file: None,
+    }
+}
+
+impl SpooledTempFile {
+    /// Returns true if the file has been rolled over to disk.
```

Comment:
Can we have a `new(size)` constructor? I know it duplicates the `spooled_tempfile` function but that matches the rest of our code.

For context, we do this because `tempfile()` *doesn't* have a constructor because it doesn't have a type. Therefore, we create convenience functions for all tempfile types (for discoverability). However, it's (IMO) still idiomatic to have a constructor.